clients/drcachesim/tools/instr_decode_cache.h

```diff
@@ -0,0 +1,172 @@
+/* **********************************************************
+ * Copyright (c) 2024 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/**
+ * instr_decode_cache.h: Library that supports caching of instruction decode
+ * information.
+ */
+
+#ifndef _INSTR_DECODE_CACHE_H_
+#define _INSTR_DECODE_CACHE_H_ 1
+
+#include "dr_api.h"
+#include "memref.h"
+
+#include <unordered_map>
+
+namespace dynamorio {
+namespace drmemtrace {
+
+/**
+ * Base class for storing instruction decode info. Users should sub-class this
+ * base class and implement set_decode_info() to derive and store the decode info
+ * they need.
+ *
+ * Derived classes must provide a default constructor to make objects
+ * corresponding to an invalid decoding.
```

Comment:
But then to query whether it's valid you have to define operator== and construct an empty object.  I see you have avoided this in all uses in this PR: your test never checks whether valid; it assume it's valid when the is_nop or is_ret is set.  In invariant_checker you have has_valid_decoding: isn't that nice to have that boolean?

I think this default-being-invalid should be removed.  I think it would be much simpler to have a base class member `bool valid_ = true` and `is_valid()`.  That's easier for users; it's already there in the use case in invariant_checker.  Then you can ask whether this is valid regardless of whether you have a pointer to the base or a derived.  It is simpler and more efficient than having to do `== MyType()`, which requires the user to define an operator==.

Reply:
I did consider adding the `valid` bool in the base-class. But how do we set it automatically when set_decode_info has been successful? It can be done I guess by having set_decode_info and set_decode_info_derived, where the former calls the latter and sets valid. Will make that change.

Reply:
Done

Reply:
> I see you have avoided this in all uses in this PR: your test never checks whether valid; 

Is there a way to construct a memref that is guaranteed to fail in decode? I tried OP_INVALID, OP_add without any args, but both fail in encoding too of course. I tried messing up the encoding[] field in the memref but hard to get an invalid encoding that way.